ibis/backends/base/sqlglot/compiler.py

```diff
@@ -0,0 +1,1300 @@
+        self.f = FuncGen()
+
+    @property
+    @abc.abstractmethod
+    def dialect(self) -> str:
+        """Backend dialect."""
+
+    @property
+    @abc.abstractmethod
+    def type_mapper(self) -> type[SqlglotType]:
+        """The type mapper for the backend."""
+
+    @abc.abstractmethod
+    def _aggregate(self, funcname, *args, where):
+        """Translate an aggregate function.
+
+        Three flavors of filtering aggregate function inputs:
+
+        1. supports filter (duckdb, postgres, others)
+           e.g.: sum(x) filter (where predicate)
+        2. use null to filter out
+           e.g.: sum(if(predicate, x, NULL))
+        3. clickhouse's ${func}If implementation, e.g.:
+           sumIf(predicate, x)
+        """
+
+    # Concrete API
+
+    def if_(self, condition, true, false: sge.Expression | None = None) -> sge.If:
+        return sge.If(
+            this=sge.convert(condition),
+            true=sge.convert(true),
+            false=false if false is None else sge.convert(false),
+        )
+
+    def cast(self, arg, to: dt.DataType) -> sge.Cast:
+        return sg.cast(sge.convert(arg), to=self.type_mapper.from_ibis(to))
+
+    def translate(self, op, *, params: Mapping[ir.Value, Any]) -> sge.Expression:
+        """Translate an ibis operation to a sqlglot expression.
+
+        Parameters
+        ----------
+        op
+            An ibis operation
+        params
+            A mapping of expressions to concrete values
+        compiler
+            An instance of SQLGlotCompiler
+        translate_rel
+            Relation node translator
+        translate_val
+            Value node translator
+
+        Returns
+        -------
+        sqlglot.expressions.Expression
+            A sqlglot expression
+        """
+
+        gen_alias_index = itertools.count()
+        quoted = self.quoted
+
+        def fn(node, _, **kwargs):
+            result = self.visit_node(node, **kwargs)
+
+            # don't alias root nodes or value ops
+            if node is op or isinstance(node, ops.Value):
+                return result
+
+            alias_index = next(gen_alias_index)
+            alias = sg.to_identifier(f"t{alias_index:d}", quoted=quoted)
+
+            try:
+                return result.subquery(alias)
+            except AttributeError:
+                return result.as_(alias, quoted=quoted)
+
+        # substitute parameters immediately to avoid having to define a
+        # ScalarParameter translation rule
+        #
+        # this lets us avoid threading `params` through every `translate_val`
+        # call only to be used in the one place it would be needed: the
+        # ScalarParameter `translate_val` rule
+        params = {
+            # remove aliases from scalar parameters
+            param.op().replace(unwrap_scalar_parameter): value
+            for param, value in (params or {}).items()
+        }
+
+        op = op.replace(
+            replace_scalar_parameter(params)
```

Comment:
We should move this rule to rewrites as well, but I can do that in a follow-up.

Reply:
It's already there, no need to move it.